File: src/main/java/com/worldscape/terrain/TerrainControlPoint.java
package com.worldscape.terrain;

import com.worldscape.terrain.TerrainType;

public class TerrainControlPoint {
    public final int x;
    public final int z;
    public final TerrainType terrainType;
    public final double elevationOffset;
    public final double influenceRadius;

    public TerrainControlPoint(int x, int z, TerrainType terrainType, double elevationOffset, double influenceRadius) {
        // 参数验证：influenceRadius 必须为正数，防止除零和无效影响范围
        // Parameter validation: influenceRadius must be positive to prevent division by zero and invalid influence ranges
        if (influenceRadius <= 0.0) throw new IllegalArgumentException("influenceRadius must be positive, got: " + influenceRadius);
        this.x = x;
        this.z = z;
        this.terrainType = terrainType;
        // 构造时钳制偏移量，确保海拔偏移在合法范围内
        // Clamp offset on construction to ensure elevation offset is within valid range
        this.elevationOffset = this.clampOffset(terrainType, elevationOffset);
        this.influenceRadius = influenceRadius;
    }

    private double clampOffset(TerrainType type, double rawOffset) {
        double minOffset = 0.0;
        double maxOffset;
        if (type == TerrainType.PLAINS) {
            maxOffset = 5.0;
            minOffset = -5.0;
        } else if (type == TerrainType.HILLS) {
            maxOffset = 30.0;
            minOffset = 10.0;
        } else if (type == TerrainType.RIDGE) {
            maxOffset = 80.0;
            minOffset = 40.0;
        } else if (type == TerrainType.HIGH_MOUNTAINS) {
            maxOffset = 150.0;
            minOffset = 80.0;
        } else if (type == TerrainType.CANYON) {
            // Deep erosional cut: modifier -50.0, energy*50-50=[-50,0] → clamps to [-50,-30]
            maxOffset = -30.0;
            minOffset = -80.0;
        } else if (type == TerrainType.DUNE) {
            maxOffset = 15.0;
            minOffset = 0.0;
        } else if (type == TerrainType.PLATEAU) {
            maxOffset = 60.0;
            minOffset = 30.0;
        } else if (type == TerrainType.DOME) {
            maxOffset = 35.0;
            minOffset = 15.0;
        } else if (type == TerrainType.CLIFF) {
            maxOffset = 0.0;
            minOffset = 0.0;
        } else if (type == TerrainType.BASIN) {
            // Broad depression: modifier -25.0, energy*50-25=[-25,25] → clamps to [-25,-10]
            maxOffset = -10.0;
            minOffset = -30.0;
        } else if (type == TerrainType.GLACIAL_VALLEY) {
            // U-shaped glacial depression: modifier -30.0, energy*50-30=[-30,20] → clamps to [-30,-10]
            maxOffset = -10.0;
            minOffset = -40.0;
        } else if (type == TerrainType.VALLEY) {
            // Moderate erosional depression: modifier -15.0, energy*50-15=[-15,35] → clamps to [-15,-5]
            maxOffset = -5.0;
            minOffset = -25.0;
        } else if (type == TerrainType.CIRQUE) {
            // Bowl-shaped glacial depression: modifier -20.0, energy*50-20=[-20,30] → clamps to [-20,-5]
            maxOffset = -5.0;
            minOffset = -30.0;
        } else if (type == TerrainType.FJORD) {
            // Deep coastal inlet: modifier -30.0, energy*50-30=[-30,20] → clamps to [-30,-10]
            maxOffset = -10.0;
            minOffset = -40.0;
        } else if (type == TerrainType.SINKHOLE) {
            // Deep narrow depression: modifier -25.0, energy*50-25=[-25,25] → clamps to [-25,-10]
            maxOffset = -10.0;
            minOffset = -30.0;
        } else if (type == TerrainType.BEACH) {
            maxOffset = 6.0;
            minOffset = -2.0;
        } else if (type == TerrainType.TRENCH) {
            maxOffset = -60.0;
            minOffset = -100.0;
        } else if (type == TerrainType.SEA_PLATEAU) {
            maxOffset = -10.0;
            minOffset = -30.0;
        } else if (type == TerrainType.DELTA) {
            maxOffset = 5.0;
            minOffset = -5.0;
        } else {
            maxOffset = 10.0;
            minOffset = -10.0;
        }
        return Math.max(minOffset, Math.min(maxOffset, rawOffset));
    }

    public double calculateInfluence(int targetX, int targetZ) {
        double dx = targetX - this.x;
        double dz = targetZ - this.z;
        double distanceSq = dx * dx + dz * dz;
        double distance = Math.sqrt(distanceSq);
        if (distance > this.influenceRadius) {
            return 0.0;
        }
        double normalizedDist = distance / this.influenceRadius;
        return (1.0 - normalizedDist) * (1.0 - normalizedDist);
    }

    public double calculateInfluenceFromSquaredDist(double distanceSq) {
        double radiusSq = this.influenceRadius * this.influenceRadius;
        if (distanceSq > radiusSq) {
            return 0.0;
        }
        double normalizedDist = Math.sqrt(distanceSq) / this.influenceRadius;
        return (1.0 - normalizedDist) * (1.0 - normalizedDist);
    }

    public double squaredDistanceTo(int targetX, int targetZ) {
        double dx = targetX - this.x;
        double dz = targetZ - this.z;
        return dx * dx + dz * dz;
    }

    public int getX() {
        return this.x;
    }

    public int getZ() {
        return this.z;
    }

    public TerrainType getTerrainType() {
        return this.terrainType;
    }

    public double getElevationOffset() {
        return this.elevationOffset;
    }

    public double getRadius() {
        return this.influenceRadius;
    }
}

File: src/main/kotlin/com/worldscape/terrain/TerrainControlPoint.kt
package com.worldscape.terrain

class TerrainControlPoint(
    @JvmField val x: Int,
    @JvmField val z: Int,
    @JvmField val terrainType: TerrainType,
    elevationOffset: Double,
    @JvmField val influenceRadius: Double
) {
    @JvmField val elevationOffset: Double

    init {
        require(influenceRadius > 0.0) { "influenceRadius must be positive, got: $influenceRadius" }
        this.elevationOffset = clampOffset(terrainType, elevationOffset)
    }

    private fun clampOffset(type: TerrainType, rawOffset: Double): Double {
        // Prefer JSON-defined offset range; fall back to legacy hardcoded mapping if no function def is loaded.
        // 优先使用 JSON 定义的偏移范围；未加载函数定义时回退到旧硬编码映射。
        val functionDef = type.getFunctionDef()
        val (minOffset, maxOffset) = if (functionDef != null) {
            functionDef.minOffset to functionDef.maxOffset
        } else {
            getLegacyOffsetRange(type)
        }
        return rawOffset.coerceIn(minOffset, maxOffset)
    }

    /**
     * Legacy hardcoded elevation offset ranges used when no JSON function definition is loaded.
     * Kept as a safe fallback to preserve behavior for any future path that constructs a
     * TerrainControlPoint before defaults.json is available.
     * 未加载 JSON 函数定义时使用的旧硬编码海拔偏移范围。保留为安全回退，以应对在 defaults.json 可用前构造 TerrainControlPoint 的未来路径。
     */
    private fun getLegacyOffsetRange(type: TerrainType): Pair<Double, Double> {
        return when (type) {
            TerrainType.PLAINS -> -5.0 to 5.0
            TerrainType.HILLS -> 10.0 to 30.0
            TerrainType.RIDGE -> 40.0 to 80.0
            TerrainType.HIGH_MOUNTAINS -> 80.0 to 150.0
            TerrainType.CANYON -> -80.0 to -30.0
            TerrainType.DUNE -> 0.0 to 15.0
            TerrainType.PLATEAU -> 30.0 to 60.0
            TerrainType.DOME -> 15.0 to 35.0
            TerrainType.CLIFF -> 0.0 to 0.0
            TerrainType.BASIN -> -30.0 to -10.0
            TerrainType.GLACIAL_VALLEY -> -40.0 to -10.0
            TerrainType.VALLEY -> -25.0 to -5.0
            TerrainType.CIRQUE -> -30.0 to -5.0
            TerrainType.FJORD -> -40.0 to -10.0
            TerrainType.SINKHOLE -> -30.0 to -10.0
            TerrainType.BEACH -> -2.0 to 6.0
            TerrainType.TRENCH -> -100.0 to -60.0
            TerrainType.SEA_PLATEAU -> -30.0 to -10.0
            TerrainType.DELTA -> -5.0 to 5.0
            else -> -10.0 to 10.0
        }
    }

    fun calculateInfluence(targetX: Int, targetZ: Int): Double {
        val dx = (targetX - x).toDouble()
        val dz = (targetZ - z).toDouble()
        val distanceSq = dx * dx + dz * dz
        val distance = Math.sqrt(distanceSq)
        if (distance > influenceRadius) return 0.0
        val normalizedDist = distance / influenceRadius
        return (1.0 - normalizedDist) * (1.0 - normalizedDist)
    }

    fun calculateInfluenceFromSquaredDist(distanceSq: Double): Double {
        val radiusSq = influenceRadius * influenceRadius
        if (distanceSq > radiusSq) return 0.0
        val normalizedDist = Math.sqrt(distanceSq) / influenceRadius
        return (1.0 - normalizedDist) * (1.0 - normalizedDist)
    }

    fun squaredDistanceTo(targetX: Int, targetZ: Int): Double {
        val dx = (targetX - x).toDouble()
        val dz = (targetZ - z).toDouble()
        return dx * dx + dz * dz
    }
}